### PR TITLE
YD-672 Solved 500 error on request overwrite code

### PR DESCRIPTION
**Pull request by Bert-R**
Concurrent requests for an account overwrite code could lead to an exception. Next to that, people could use this API to spam users and drain money from Yona.

Two mechanisms are implemented:
* If an account overwrite code is requested for a given number, subsequent requests for that same number will be ignored for a configurable duration (default 30 seconds)
* The DOSProtectionService is enabled for the "requestOverwriteUserConfirmationCode" API. After more than 3 requests within 5 minutes, a delay is introduced. This is independent of the mobile number used.

### CODE REVIEW

File: appservice/src/main/java/nu/yona/server/subscriptions/rest/PinResetRequestController.java [hunk 1]
```diff
@@ -132,13 +132,13 @@
...
-		ConfirmationCode confirmationCode = userService.getUserEntityById(userResource.getContent().getId())
+		Optional<ConfirmationCode> confirmationCode = userService.getUserEntityById(userResource.getContent().getId())
 				.getPinResetConfirmationCode();
-		if (confirmationCode == null || pinResetRequestService.isExpired(confirmationCode))
+		if (!pinResetRequestService.isValidConfirmationCode(confirmationCode))
 		{
 			addPinResetRequestLink(userResource);
 		}
-		else if (confirmationCode.getCode() != null)
+		else if (confirmationCode.get().getCode() != null)
```
Comment (yonabuild) [inline appservice/src/main/java/nu/yona/server/subscriptions/rest/PinResetRequestController.java]: MAJOR Call "confirmationCode.isPresent()" before accessing the value. [rule](http://10.127.10.107:9000/coding_rules#rule_key=squid%3AS3655)

File: core/src/main/java/nu/yona/server/subscriptions/service/UserUpdateService.java [hunk 2]
```diff
@@ -1,9 +1,10 @@
 /*******************************************************************************
- * Copyright (c) 2019 Stichting Yona Foundation This Source Code Form is subject to the terms of the Mozilla Public License, v.
+ * Copyright (c) 2019, 2020 Stichting Yona Foundation This Source Code Form is subject to the terms of the Mozilla Public License, v.
  * 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
  *******************************************************************************/
 package nu.yona.server.subscriptions.service;
 
+import java.time.Duration;
```
Comment (yonabuild) [inline core/src/main/java/nu/yona/server/subscriptions/service/UserUpdateService.java]: MINOR Remove this unused import 'java.time.Duration'. [rule](http://10.127.10.107:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: core/src/main/java/nu/yona/server/subscriptions/service/PinResetRequestService.java [hunk 3]
```diff
@@ -80,9 +82,10 @@ public void verifyPinResetConfirmationCode(UUID userId, String userProvidedConfi
 		User userEntity = userService.getUserEntityById(userId);
 		logger.info("User with mobile number '{}' and ID '{}' requested to verify the pin reset confirmation code",
 				userEntity.getMobileNumber(), userId);
-		ConfirmationCode confirmationCode = userEntity.getPinResetConfirmationCode();
-		Require.that((confirmationCode != null) && !isExpired(confirmationCode),
+		Optional<ConfirmationCode> confirmationCodeOpt = userEntity.getPinResetConfirmationCode();
+		Require.that(isValidConfirmationCode(confirmationCodeOpt),
 				() -> PinResetRequestConfirmationException.confirmationCodeNotSet(userEntity.getMobileNumber()));
+		ConfirmationCode confirmationCode = confirmationCodeOpt.get();
```
Comment (yonabuild) [inline core/src/main/java/nu/yona/server/subscriptions/service/PinResetRequestService.java]: MAJOR Call "confirmationCodeOpt.isPresent()" before accessing the value. [rule](http://10.127.10.107:9000/coding_rules#rule_key=squid%3AS3655)

File: appservice/src/main/java/nu/yona/server/admin/rest/AdminController.java [hunk 4]
```diff
@@ -1,14 +1,19 @@
...
 package nu.yona.server.admin.rest;
 
+import nu.yona.server.DOSProtectionService;
+import nu.yona.server.properties.YonaProperties;
+import nu.yona.server.rest.ControllerBase;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder;
 import org.springframework.http.HttpStatus;
```
Comment (yonabuild) [inline appservice/src/main/java/nu/yona/server/admin/rest/AdminController.java]: MINOR Remove this unused import 'org.springframework.http.HttpStatus'. [rule](http://10.127.10.107:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: core/src/main/java/nu/yona/server/subscriptions/service/UserAddService.java [hunk 5]
```diff
@@ -262,13 +262,21 @@ UserDto addUserCreatedOnBuddyRequest(UserDto buddyUserResource)
...
 			Supplier<YonaException> tooManyAttemptsExceptionSupplier)
 	{
-		Require.isNonNull(confirmationCode, noConfirmationCodeExceptionSupplier);
+		assertValidConfirmationCode(userEntity, confirmationCode.orElseThrow(noConfirmationCodeExceptionSupplier),
+				userProvidedConfirmationCode, noConfirmationCodeExceptionSupplier, invalidConfirmationCodeExceptionSupplier,
+				tooManyAttemptsExceptionSupplier);
+	}
 
+	private void assertValidConfirmationCode(User userEntity, ConfirmationCode confirmationCode, String userProvidedConfirmationCode,
+			Supplier<YonaException> noConfirmationCodeExceptionSupplier,
```
Comment (yonabuild) [inline core/src/main/java/nu/yona/server/subscriptions/service/UserAddService.java]: MAJOR Remove this unused method parameter "noConfirmationCodeExceptionSupplier". [rule](http://10.127.10.107:9000/coding_rules#rule_key=squid%3AS1172)

**Comment by yonabuild** [comment]
SonarQube analysis reported 6 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [AdminController.java#L21](https://github.com/yonadev/yona-server/blob/18c6877854f838c310cd5e82a99ebd25d2d7ec85/appservice/src/main/java/nu/yona/server/admin/rest/AdminController.java#L21): Remove this unused import 'org.springframework.web.bind.annotation.ResponseStatus'. [rule](http://10.127.10.107:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: core/src/main/java/nu/yona/server/subscriptions/service/PinResetRequestService.java (hunk 3 above)
Comment (Bert-R) [reply core/src/main/java/nu/yona/server/subscriptions/service/PinResetRequestService.java]: False positive. The check on line 86 verifies that the confirmation code is present.

File: appservice/src/main/java/nu/yona/server/subscriptions/rest/PinResetRequestController.java (hunk 1 above)
Comment (Bert-R) [reply appservice/src/main/java/nu/yona/server/subscriptions/rest/PinResetRequestController.java]: False positive. The check on line 137 verifies that the confirmation code is present.

File: core/src/main/java/nu/yona/server/subscriptions/service/UserAddService.java [hunk 6]
```diff
@@ -262,13 +262,21 @@ UserDto addUserCreatedOnBuddyRequest(UserDto buddyUserResource)
...
 			Supplier<YonaException> tooManyAttemptsExceptionSupplier)
 	{
-		Require.isNonNull(confirmationCode, noConfirmationCodeExceptionSupplier);
+		assertValidConfirmationCode(userEntity, confirmationCode.orElseThrow(noConfirmationCodeExceptionSupplier),
+				userProvidedConfirmationCode, noConfirmationCodeExceptionSupplier, invalidConfirmationCodeExceptionSupplier,
+				tooManyAttemptsExceptionSupplier);
+	}
 
+	private void assertValidConfirmationCode(User userEntity, ConfirmationCode confirmationCode,
+			String userProvidedConfirmationCode, Supplier<YonaException> noConfirmationCodeExceptionSupplier,
```
Comment (yonabuild) [inline core/src/main/java/nu/yona/server/subscriptions/service/UserAddService.java]: MAJOR Remove this unused method parameter "noConfirmationCodeExceptionSupplier". [rule](http://10.127.10.107:9000/coding_rules#rule_key=squid%3AS1172)

**Comment by yonabuild** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by yonabuild** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: core/src/main/java/nu/yona/server/properties/YonaProperties.java [hunk 7]
```diff
@@ -191,4 +194,14 @@ public void setTestServer(boolean isTestServer)
 	{
 		this.isTestServer = isTestServer;
 	}
+
+	public Duration getOverwriteUserConfirmationCodeValidityTime()
```
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/properties/YonaProperties.java:198]: The name seems to overlap with the existing `getPinResetRequestExpirationTime` property. "Valid" and "not expired" seem to mean the same. So shouldn't we just add the rate limiting/throttling and for the rest keep sending, even within 30 seconds?

**Review by aristotelos — COMMENTED** [review]
Looks good in general. Suggestion: We may have a look at existing rate limit/throttle solutions available with Spring, such as [Bucket4j](https://www.baeldung.com/spring-bucket4j) or [spring-boot-throttling](https://medium.com/teamarimac/implementing-throttling-in-java-spring-boot-ec4723cfce9f) instead of implementing our own, although it is quite simple. I like the `@Throttling` annotation of spring-boot-throttling.

Added another single question.

File: core/src/main/java/nu/yona/server/properties/YonaProperties.java (hunk 7 above)
Comment (Bert-R) [reply core/src/main/java/nu/yona/server/properties/YonaProperties.java:211]: The app has the unpleasant habit of sending requests multiple times in quick succession. Given only a few requests are made, throttling does not address it, but we're sending multiple SMSes (extra cost) and we get concurrency issues.
However, I agree the name isn't great. It sounds like expiration time, but it isn't.  It's the "nonresend" or "je doet het er maar mee" interval. What would be a good name for that?

**Comment by Bert-R** [comment]
Pushed an update for the rename.
For the throttling suggestion, I've logged [YD-693](https://jira.yona.nu/browse/YD-693).

File: core/src/main/java/nu/yona/server/properties/YonaProperties.java (hunk 7 above)
Comment (aristotelos) [reply core/src/main/java/nu/yona/server/properties/YonaProperties.java]: The new name looks great! 😄 

"The app has the unpleasant habit of sending requests multiple times in quick succession." Shouldn't we fix that habit instead? I think that is not unreasonable?

**Review by aristotelos — APPROVED** [review]

File: core/src/main/java/nu/yona/server/properties/YonaProperties.java (hunk 7 above)
Comment (Bert-R) [reply core/src/main/java/nu/yona/server/properties/YonaProperties.java]: Absolutely, but I've asked that before, with no result yet. App development doesn't make much progress at the moment.

**Pull request merged**